### DISCUSSION TITLE
Missing OSM Restrictions in Bounding Box Query After Importing with osm2pgsql

### DISCUSSION THREAD

**Discussion by ghost** [osm2pgsql-dev/osm2pgsql#2215] [Q&A]
Hi I’m using osm2pgsql to import OpenStreetMap (OSM) data into a PostgreSQL database for analysis. 
Sadly I’m not seeing all the expected turn restrictions within a specified bounding box.

**osm2pgsql Command:**

`osm2pgsql -U postgres -W -d osm_data -H localhost -s -G --number-processes 30 -C 102400 dach-latest.osm.pbf
`

**Modifications to default.style:**

```
node,way   restriction                    text         linear
node,way   restriction:conditional        text         linear
```

**SQL Query:**

```
SELECT DISTINCT restriction, "restriction:conditional"
FROM planet_osm_line
WHERE highway IN ('motorway', 'trunk', 'primary', 'secondary', 'tertiary', 'unclassified',
                  'residential', 'motorway_link', 'trunk_link', 'primary_link', 'secondary_link', 'tertiary_link')
AND ST_Intersects(way, ST_Transform(ST_MakeEnvelope(10.91492, 47.41136, 13.41705, 49.07207, 4326), 3857));

```

```
restriction   |              restriction:conditional               
-----------------+----------------------------------------------------
 only_right_turn | 
                 | restriction:conditional=no_parking @ (19:00-06:00)
                 | 
(3 rows)
```

I have verified the presence of more restrictions within this bounding box using  https://ahorn.lima-city.de/tr/ 
which shows a greater variety of restrictions.

**What I’ve Tried:**

Updating Data: Ensured my database is up-to-date with the latest OSM data.
Query Adjustments: Simplified and expanded the query to include all potential restrictions without the highway filter.

**Question:**

What could be causing the discrepancy between the restrictions visible in the external tool and those retrieved from my PostgreSQL database? 

Are there any additional configurations or steps I should consider to ensure all relevant restrictions are captured?

Any insights or suggestions would be greatly appreciated!

Thank you!

**Comment by joto** [comment]
Turn restrictions are mapped as special relations. They can not just be mapped to simple lines, processing will be much more involved. The old pgsql output of osm2pgsql was never intended to handle them and will simply not work for you. The newer flex output of osm2pgsql can, in theory, handle turn restriction relations, but some configuration and post-processing will be necessary to make the data useful. Please familiarize yourself with how turn restriction relations work and with the flex output. You might be able to get something to work, but it will need some effort. I have not yet seen a configuration for osm2pgsql that handles turn restrictions.

**Reply by ghost** [reply]
> Turn restrictions are mapped as special relations. They can not just be mapped to simple lines, processing will be much more involved. The old pgsql output of osm2pgsql was never intended to handle them and will simply not work for you. The newer flex output of osm2pgsql can, in theory, handle turn restriction relations, but some configuration and post-processing will be necessary to make the data useful. Please familiarize yourself with how turn restriction relations work and with the flex output. You might be able to get something to work, but it will need some effort. I have not yet seen a configuration for osm2pgsql that handles turn restrictions.

I modified a lua script I found here : 

https://raw.githubusercontent.com/osm2pgsql-dev/osm2pgsql/master/flex-config/compatible.lua

to include restrictions:

https://raw.githubusercontent.com/Andreas4242/Andreas_file_tools/main/compatible.lua

Im not sure its totally correct but it gives me results if I run this query:

```
SELECT DISTINCT tr.restriction, tr.tags->>'restriction:conditional'
FROM planet_osm_line l
JOIN planet_osmturn_restrictions tr
ON l.osm_id = tr.from_way OR l.osm_id = tr.to_way
WHERE l.highway IN ('motorway', 'trunk', 'primary', 'secondary', 'tertiary', 'unclassified',
                    'residential', 'motorway_link', 'trunk_link', 'primary_link', 'secondary_link', 'tertiary_link')
AND ST_Intersects(l.way, ST_Transform(ST_MakeEnvelope(10.91492, 47.41136, 13.41705, 49.07207, 4326), 3857));

```

```
  restriction    |               ?column?                
------------------+---------------------------------------
 no_left_turn     | none @ destination
 no_left_turn     | no @ psv
 no_left_turn     | 
 no_right_turn    | 
 no_straight_on   | 
 no_turn          | no_left_turn @ (Mo-Fr 16:00-19:00)
 no_turn          | only_right_turn @ (Mo-Fr 07:00-14:00)
 no_turn          | 
 no_u_turn        | 
 only_left_turn   | 
 only_right_turn  | 
 only_straight_on | 
 only_u_turn      | 
(13 rows)
```

This whole lua business is waaaayy to complicated for me.. loved the old config more..  im only half sure  what im doing and thank god for chatgpt :D
